Document how operator hierarchy is merged. (#2426)
grafana/agent: docs/sources/operator/architecture.md

GrafanaAgents use the same Probe, modifying that Probe will cause both
GrafanaAgents to be reconciled.

### Build the Hierarchy

The hierarchy is built via label matching on the custom resources. Below is a
graph illustrating the matching. The `GrafanaAgent` picks up the `MetricsInstance`

> and `LogsInstance` that match the labels `app.kubernetes.io/name: loki` and `app.kubernetes.io/instance: release-name`. The instances pick up the resources the same way.

Author: We probably should use different labels for an example.

Reviewer: What would be good labels for the example?  Might want to do it now because it might slip through the cracks...

Author: I believe the pattern is very common. Unfortunately I don't know the space well enough.

Reviewer: @captncraig Do you know what labels we could use instead?

Reviewer 2: In the quickstart we use:

```
  metrics:
    instanceSelector:
      matchLabels:
        agent: grafana-agent-metrics
  logs:
    instanceSelector:
      matchLabels:
        agent: grafana-agent-logs
```

and 

```
podMonitorSelector:
    matchLabels:
      instance: primary
```

I'd usually recommend using the `name` of the parent object as the label value. And it looks like we have been using `instance` as the key. Maybe we should standardize on a more namespaced label key, but thats a different discussion entirely.

> The hierarchy is built via label matching on the custom resources. Below is a
> graph illustrating the matching. The `GrafanaAgent` picks up the `MetricsInstance`
> and `LogsInstance` that match the labels `app.kubernetes.io/name: loki` and `app.kubernetes.io/instance: release-name`. The instances pick up the resources the same way.

Reviewer: ```suggestion
Grafana Agent Operator builds the hierarchy using label matching on the custom resources. The following figure illustrates the matching. The `GrafanaAgent` picks up the `MetricsInstance`
and `LogsInstance` that match the labels `app.kubernetes.io/name: loki` and `app.kubernetes.io/instance: release-name`, respectively. The instances pick up the resources the same way.
```

> <img src="../../assets/hierarchy.svg" alt="Hierarchy" />

Reviewer: The text on the edges in this graph is always the same and it makes the image busy.  Should we leave it out since you say this in the text above the image?  Another approach might be to add a callout on the left that points to some of the edges that says:

"The operator matches each instance against:

```matchLabels
   app.kubernetes.io/name: loki
   app.kubernetes.io/instance: release-name
```

Reviewer: I can work on it later because I'm working on these topics anyways.  It was created by a former writer, Karen M (not to be confused with me, Karen G. 😄 ) .

Reviewer 2: I am also confused about the match labels. The diagram just has `name:` not the full `app.kubernetes.io/name:`. I'd probably update the diagram with the full names?

### Debug the Hierarchy

> The generated configurations are save in secrets. They can be downloaded and
> validated manually with the following commands.

Reviewer: ```suggestion
The generated configurations are saved in secrets. To download and
validate them manually, use the following commands:
```

```
$ kubectl get secrets <???>-logs-config -o json | jq -r '.data."agent.yml"' | base64 --decode

> $ kubectl get secrets <???>-config -o json | jq -r '.data."agent.yml"' | base64 --decode

Author: How are the config names defined? What is the pattern?
